Add validator check for `treeSelect`

GinnyJI opened the pull request:
We check if the raw string input from user is a valid DNS label. If this
is a valid DNS label but couldn't be parsed as a selector for some
reason, we will show user that it's an internal error, so that user can
reach us for help. In theory this should never happen.

Tested:
```
$k delete role test -n a
k create role test -n a --verb=update --resource=deployments
k annotate role test -n a "propagate.hnc.x-k8s.io/treeSelect"="1234567890123456789012345678901234567890123456789012345678901234567890"

role.rbac.authorization.k8s.io "test" deleted
role.rbac.authorization.k8s.io/test created
Error from server (BadRequest): admission webhook "objects.hnc.x-k8s.io" denied the request: Invalid HNC "propagate.hnc.x-k8s.io/treeSelect" value: 1234567890123456789012345678901234567890123456789012345678901234567890 is not a valid namespace name, must be no more than 63 characters
```

GinnyJI commented:
/assign @adrianludwin 
/assign @yiqigao217

adrianludwin commented:
That error message doesn't look right to me:

```
The request is invalid: : Invalid Kubernetes labelSelector: 
1234567890123456789012345678901234567890123456789012345678901234567890,
must be no more than 63 characters
```

There's an extra colon in there for some reason, but more importantly, this isn't a K8s labelSelector, it's an HNC tree selector with very different rules. We're not giving users the information that the problem is that the tree selector contains invalid namespace names.

GinnyJI commented:
> That error message doesn't look right to me:
> 
> ```
> The request is invalid: : Invalid Kubernetes labelSelector: 
> 1234567890123456789012345678901234567890123456789012345678901234567890,
> must be no more than 63 characters
> ```
> 
> There's an extra colon in there for some reason, but more importantly, this isn't a K8s labelSelector, it's an HNC tree selector with very different rules. We're not giving users the information that the problem is that the tree selector contains invalid namespace names.

Updated the error msg. I looked into the double colons, I don't _think_ it's from my change. I haven't found out the place that's causing this, still looking

adrianludwin commented:
> Updated the error msg. I looked into the double colons, I don't _think_ it's from my change. I haven't found out the place that's causing this, still looking

Out of interest, if you change the response code (e.g. from `metav1.StatusReasonInvalid` to something else, like `metav1.StatusReasonBadRequest`)? I agree that the code you picked is the correct one, I'm just wondering if there's something funny about the way it gets displayed to the user.

adrianludwin commented:
Also you need to identify the `treeSelect` annotation somewhere or users won't know which field is causing the problem.

adrianludwin commented on incubator/hnc/internal/validators/object.go:
```diff
@@ -142,6 +148,24 @@ func validateSelectorChange(inst, oldInst *unstructured.Unstructured) error {
+	if newSelectorStr == "" || oldSelectorStr == newSelectorStr {
+		return nil
+	}
+	errStrs := validation.IsDNS1123Label(newSelectorStr)
+	if len(errStrs) != 0 {
+		return errors.New(newSelectorStr + ", " + strings.Join(errStrs, ". "))
+	}
+	_, err := selectors.GetTreeSelector(inst)
```
The error message should be generated in the `selectors` package, not here, so that we can display the _same_ error from the reconciler if needed. `GetTreeSelector` should be the one that returns the error.

GinnyJI replied:
done

GinnyJI commented:
> > Updated the error msg. I looked into the double colons, I don't _think_ it's from my change. I haven't found out the place that's causing this, still looking
> 
> Out of interest, if you change the response code (e.g. from `metav1.StatusReasonInvalid` to something else, like `metav1.StatusReasonBadRequest`)? I agree that the code you picked is the correct one, I'm just wondering if there's something funny about the way it gets displayed to the user.

Yeah if we use `metav1.StatusReasonBadRequest` the error msg looks better formatted
```
Error from server (BadRequest): admission webhook "objects.hnc.x-k8s.io" denied the request: Invalid HNC treeSelect value: 1234567890123456789012345678901234567890123456789012345678901234567890, must be no more than 63 characters
```

adrianludwin commented:
> Yeah if we use `metav1.StatusReasonBadRequest` the error msg looks better formatted

Weird. I agree that `Invalid` is a better error code here, but maybe for now just use `BadRequest` and file a bug to look into how to change it back to `Invalid`? I suspect that there's something we can add to the webhook status message that we return that would format it better, but it's probably not worth looking into right now.

GinnyJI commented:
> Also you need to identify the `treeSelect` annotation somewhere or users won't know which field is causing the problem.

done

GinnyJI commented:
> > Yeah if we use `metav1.StatusReasonBadRequest` the error msg looks better formatted
> 
> Weird. I agree that `Invalid` is a better error code here, but maybe for now just use `BadRequest` and file a bug to look into how to change it back to `Invalid`? I suspect that there's something we can add to the webhook status message that we return that would format it better, but it's probably not worth looking into right now.

Changed both of the errors for `select` and `treeSelect`, and filed the bug at https://github.com/kubernetes-sigs/multi-tenancy/issues/1218

adrianludwin commented on incubator/hnc/internal/pkg/selectors/selectors.go:
```diff
@@ -18,18 +20,28 @@ func GetSelectorAnnotation(inst *unstructured.Unstructured) string {
 		return nil, nil
 	}
+	// check if it's a valid namespace name
+	errStrs := validation.IsDNS1123Label(treeSelectorStr)
```
`a,b` is a valid tree selector string but is not a valid DNS label

GinnyJI replied:
Done. The e2e test also caught this

GinnyJI commented:
tracked at #1123

adrianludwin commented:
Two comments about the latest error message (newlines added):

```
Error from server (BadRequest): admission webhook "objects.hnc.x-k8s.io" denied the request:
Invalid HNC treeSelect value:
1234567890123456789012345678901234567890123456789012345678901234567890
is not a valid namespace name, must be no more than 63 characters

```

1. Nothing here points the user to the annotation that caused the problem
2. What does this message look like if there are multiple validation errors? (Have you tested this?)

adrianludwin commented on incubator/hnc/internal/pkg/selectors/selectors.go:
```diff
@@ -18,29 +20,47 @@ func GetSelectorAnnotation(inst *unstructured.Unstructured) string {
+			errStrs = validation.IsDNS1123Label(str)
+		}
+		if len(errStrs) != 0 {
+			return nil, errors.New(str + " is not a valid namespace name, " + strings.Join(errStrs, ". "))
```
`str` here could include the exclamation mark, because you didn't actually strip it out on line 46. Did you want to include the exclamation mark here?

GinnyJI replied:
done

adrianludwin commented:
This change will also need tests in `internal/validators` before it goes in

GinnyJI commented:
> Two comments about the latest error message (newlines added):
> 
> ```
> Error from server (BadRequest): admission webhook "objects.hnc.x-k8s.io" denied the request:
> Invalid HNC treeSelect value:
> 1234567890123456789012345678901234567890123456789012345678901234567890
> is not a valid namespace name, must be no more than 63 characters
> ```
> 
> 1. Nothing here points the user to the annotation that caused the problem
> 2. What does this message look like if there are multiple validation errors? (Have you tested this?)

Forgot to update the PR msg... updated now and it contains `treeSelect`

I wrote a brief comment in the code on what it will look like with multiple errors

adrianludwin commented:
> Forgot to update the PR msg... updated now and it contains `treeSelect`

Remember that one of your use cases is a user who's applying a YAML file they didn't write via `kubectl apply -f foo.yaml` and doesn't know anything about HNC. How would they know where to look if they just see the word "treeSelect"?

adrianludwin commented on incubator/hnc/internal/pkg/selectors/selectors.go:
```diff
@@ -18,33 +19,62 @@ func GetSelectorAnnotation(inst *unstructured.Unstructured) string {
-		return nil, fmt.Errorf("while parsing %q: %w", api.AnnotationTreeSelector, err)
+		// In theory this should never happen because we already checked DNS label before.
+		// If this happens, it's more likely that we have a bug in our code
+		return nil, fmt.Errorf("Internal error while parsing %q: %w", api.AnnotationTreeSelector, err)
```
Error usually start with a lowercase letter since they'll often get wrapped, like you did here :)

Log message, by contrast, start with upper case letters because they can't be wrapped (same with user messages).

GinnyJI replied:
Good to know! Corrected

adrianludwin commented on incubator/hnc/internal/pkg/selectors/selectors.go:
```diff
@@ -18,33 +19,62 @@ func GetSelectorAnnotation(inst *unstructured.Unstructured) string {
 	return treeSelector, nil
 }
 
+func validateNamespaceName(seg string) error {
```
You're not really passing a namespace to this function, you're passing a treeSelector segment. So maybe `validateTreeSelectorSegment`?

GinnyJI replied:
Done

adrianludwin commented on incubator/hnc/internal/pkg/selectors/selectors.go:
```diff
@@ -18,33 +19,62 @@ func GetSelectorAnnotation(inst *unstructured.Unstructured) string {
+	if len(errStrs) != 0 {
+		// If IsDNS1123Label() returns multiple errors, it will look like:
+		// ""ns" is not a valid namespace name, err1; err2
+		return fmt.Errorf("%q is not a valid namespace name, %s", ns, strings.Join(errStrs, "; "))
```
> ""ns" is not a valid namespace name, err1; err2

I assume that first double quote in the comment is just a typo?

This would look better IMO with a full colon (:), which is used to create the start of a list:

`"ns" is not a valid namespace name: error1; error 2`

GinnyJI replied:
Done. Yea it was a typo

adrianludwin commented on incubator/hnc/internal/validators/object_test.go:
```diff
@@ -146,6 +146,7 @@ func TestUserChanges(t *testing.T) {
 	f := forest.NewForest()
 	o := &Object{Forest: f}
 	l := zap.Logger(false)
+	invalidSelStr := "1234567890123456789012345678901234567890123456789012345678901234567890"
```
This only gets used once so I'd recommend inlining it into the test.

GinnyJI replied:
I'm thinking of using it in the selector test as well

adrianludwin replied:
Ah, I see. I think it might be better if each test had its own, different, bad selector :) E.g. "a b c" is a bad namespace name _and_ a bad selector.

Constants like this make the tests harder to read, because you can't just look at the test definition to know what's being tested. Constants can add clarity in same cases (e.g. you don't really have to look inside RBACGroup to know what's there) but "invalidSelector" could have anything in there, so it makes the test a bit harder to read IMO.

I know I sometimes say "please use constants!" and it's weird that I'm suggesting _not_ using them here, but I hope that explains why!

GinnyJI replied:
I see what you mean. Yeah I agree, changed the test string here

adrianludwin commented on incubator/hnc/internal/validators/object_test.go:
```diff
@@ -308,6 +309,20 @@ func TestUserChanges(t *testing.T) {
 			},
 		},
+	}, {
+		name: "Deny creation of object with invalid treeSelect annotation",
```
Probably also a good idea to have an "allow" test.

GinnyJI replied:
Added

GinnyJI commented:
> > Forgot to update the PR msg... updated now and it contains `treeSelect`
> 
> Remember that one of your use cases is a user who's applying a YAML file they didn't write via `kubectl apply -f foo.yaml` and doesn't know anything about HNC. How would they know where to look if they just see the word "treeSelect"?

ah I see, updated to use the full key

adrianludwin commented on incubator/hnc/internal/validators/object_test.go:
```diff
@@ -308,6 +308,79 @@ func TestUserChanges(t *testing.T) {
+			},
+		},
+	}, {
+		name: "Allow creation of object with valid selector and treeSelect annotation",
```
I thought we were going to disallow this? But if we haven't implemented it yet, that's fine for this PR

GinnyJI replied:
Ah right. When it's implemented I'll update this test

yiqigao217 replied:
Just confirming that we decided to only allow one type of selector at a time and cannot mix any two, right?

adrianludwin replied:
Yup!

adrianludwin reviewed (commented):
Nice work!

/lgtm
/approve
/hold
/assign @rjbez17

rjbez17 commented:
/lgtm
/hold cancel
